File: clients/typescript/src/retry.ts
import type { InboxMessage } from "./message.ts";

/** What the library does with a message whose handler threw. */
export type InboxFailureAction =
  | { readonly retry: true; readonly delayMs: number }
  | { readonly retry: false };

/** Return the message to pending after a delay. */
export function retryAfter(delayMs: number): InboxFailureAction {
  if (!Number.isFinite(delayMs) || delayMs < 0) {
    throw new RangeError("The delay must not be negative.");
  }

  return { retry: true, delayMs: Math.round(delayMs) };
}

/** Move the message to the dead letter and stop. */
export function deadLetter(): InboxFailureAction {
  return { retry: false };
}

/**
 * Decides what happens to a message whose handler threw.
 *
 * The library exposes the decision, because only the application knows that a validation error
 * must never be retried while a timeout must.
 */
export type InboxRetryPolicy = (message: InboxMessage, failure: unknown) => InboxFailureAction;

/** The settings of the default policy. */
export interface DefaultRetryPolicyOptions {
  /** The attempt ceiling. The row starts at zero, so the ceiling allows one more delivery. */
  readonly maxAttempts?: number;
  /** The delay after the first failure. */
  readonly baseDelayMs?: number;
  /** The ceiling of the backoff. */
  readonly maxDelayMs?: number;
  /** The fraction of the delay that varies, from zero to one. */
  readonly jitter?: number;
}

/**
 * Retry while `attempt < maxAttempts`, with an exponential backoff and jitter, and dead-letter
 * after that.
 *
 * @param options the settings
 * @returns the policy
 */
export function defaultRetryPolicy(options: DefaultRetryPolicyOptions = {}): InboxRetryPolicy {
  const maxAttempts = options.maxAttempts ?? 5;
  const baseDelayMs = options.baseDelayMs ?? 1000;
  const maxDelayMs = options.maxDelayMs ?? 300_000;
  const jitter = options.jitter ?? 0.2;

  if (!Number.isInteger(maxAttempts) && maxAttempts !== Number.MAX_SAFE_INTEGER) {
    throw new RangeError("The maxAttempts must be a whole number.");
  }

  if (maxAttempts <= 0) {
    throw new RangeError("The maxAttempts must be positive.");
  }

  if (baseDelayMs < 0) {
    throw new RangeError("The baseDelayMs must not be negative.");
  }

  if (maxDelayMs < baseDelayMs) {
    throw new RangeError("The maxDelayMs must not be below the baseDelayMs.");
  }

  if (jitter < 0 || jitter > 1) {
    throw new RangeError("The jitter must be between zero and one.");
  }

  return (message) => {
    if (message.attempt >= maxAttempts) {
      return deadLetter();
    }

    // A large attempt must not overflow the shift, so the exponent stops where the ceiling is
    // certainly reached.
    const exponent = Math.min(message.attempt, 30);
    const scaled = baseDelayMs * 2 ** exponent;
    const bounded = Math.min(scaled, maxDelayMs);

    if (jitter <= 0) {
      return retryAfter(bounded);
    }

    // A full band around the delay: one worker's backoff never lines up with another's.
    const offset = (Math.random() * 2 - 1) * bounded * jitter;

    return retryAfter(Math.max(0, bounded + offset));
  };
}

File: clients/go/retry.go
package queuebox

import (
	"math"
	"math/rand/v2"
	"time"
)

// FailureAction states what the library does with a message whose handler returned an error.
type FailureAction struct {
	// Retry is true to run the retry statement, and false to run the dead-letter statement.
	Retry bool
	// Delay is how long the retry waits before the message can be claimed again.
	Delay time.Duration
}

// RetryAfter returns the message to pending after a delay.
func RetryAfter(delay time.Duration) FailureAction {
	if delay < 0 {
		delay = 0
	}

	return FailureAction{Retry: true, Delay: delay}
}

// DeadLetter moves the message to the dead letter and stops.
func DeadLetter() FailureAction {
	return FailureAction{}
}

// RetryPolicy decides what happens to a message whose handler returned an error.
//
// The library exposes the decision, because only the application knows that a validation error
// must never be retried while a timeout must.
type RetryPolicy interface {
	Decide(message Message, failure error) FailureAction
}

// RetryPolicyFunc lets a plain function be a RetryPolicy.
type RetryPolicyFunc func(message Message, failure error) FailureAction

// Decide calls the function.
func (f RetryPolicyFunc) Decide(message Message, failure error) FailureAction {
	return f(message, failure)
}

type defaultRetryPolicy struct {
	maxAttempts int
	baseDelay   time.Duration
	maxDelay    time.Duration
	jitter      float64
}

// NewDefaultRetryPolicy retries while the attempt is below the ceiling, with an exponential
// backoff and jitter, and dead-letters after that.
//
// The row starts at attempt zero, so the ceiling allows maxAttempts plus one deliveries.
func NewDefaultRetryPolicy(maxAttempts int, baseDelay, maxDelay time.Duration, jitter float64) RetryPolicy {
	return defaultRetryPolicy{maxAttempts: maxAttempts, baseDelay: baseDelay, maxDelay: maxDelay, jitter: jitter}
}

func (p defaultRetryPolicy) Decide(message Message, _ error) FailureAction {
	if message.Attempt >= p.maxAttempts {
		return DeadLetter()
	}

	// A large attempt must not overflow the shift, so the exponent stops where the ceiling is
	// certainly reached.
	exponent := min(message.Attempt, 30)
	scaled := float64(p.baseDelay) * math.Pow(2, float64(exponent))
	bounded := time.Duration(math.Min(scaled, float64(p.maxDelay)))

	if p.jitter <= 0 {
		return RetryAfter(bounded)
	}

	// A full band around the delay: one worker's backoff never lines up with another's.
	offset := (rand.Float64()*2 - 1) * p.jitter * float64(bounded)

	return RetryAfter(bounded + time.Duration(offset))
}
